052-k8s-server-side-apply.md

```diff
@@ -0,0 +1,160 @@
+Modify [this line](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNamespacedResourceOperator.java#LL263C88-L263C102).
+From:
+
+```java
+....patch(PatchContext.of(PatchType.JSON), desired);
+```
+
+To:
+
+<!-- Note: change the below examples to serverSideApply once we know how -->
+```java
+....patch(new PatchContext.Builder().withPatchType(PatchType.SERVER_SIDE_APPLY).withFieldManager("strimzi-cluster-operator").withForce(true).build(), desired);
+```
+
+And to set without `Force`, as aforementioned for the first pass:
+
+```java
+....patch(new PatchContext.Builder().withPatchType(PatchType.SERVER_SIDE_APPLY).withFieldManager("strimzi-cluster-operator").withForce(false).build(), desired);
+```
+
+Also, the `StrimziPodSetOperator` patches this method [here](https://github.com/strimzi/strimzi-kafka-operator/blob/main/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StrimziPodSetOperator.java#L65), this would need updating too. Every create/update call from the existing Operator should become a `serverSideApply` patch call.
+
+And [Kubernetes](https://kubernetes.io/docs/reference/using-api/server-side-apply/#:~:text=through%20declarative%20configurations.-,Clients%20can%20create,-and%20modify%20their)/[Fabric8io](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md#server-side-apply:~:text=For%20any%20create%20or%20update) recommend it is used on CREATE also:
+[From](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNamespacedResourceOperator.java#L272):
+
+```java
+ReconcileResult<T> result = ReconcileResult.created(operation().inNamespace(namespace).resource(desired).create());
+```
+
+To:
+
+```java
+ReconcileResult<T> result = ReconcileResult.patched(operation().inNamespace(namespace).withName(name).patch(new PatchContext.Builder().withPatchType(PatchType.SERVER_SIDE_APPLY).withFieldManager("strimzi-cluster-operator").withForce(true).build(), desired);
+```
+
+And to set without `Force`, as aforementioned for the first pass:
+
+```java
+ReconcileResult<T> result = ReconcileResult.patched(operation().inNamespace(namespace).withName(name).patch(new PatchContext.Builder().withPatchType(PatchType.SERVER_SIDE_APPLY).withFieldManager("strimzi-cluster-operator").withForce(false).build(), desired);
+```
+
+Currently `ReconcileResult.patched` is used for `patch` and `ReconcileResult.created` is used for `create` operations. We will no longer be making an initial GET request to determine the current resource, so will always be using a PATCH request and returning `ReconcileResult.patched`. There is no code specifically using the `ReconcileResult.Created` type to determine behaviour that we can see, it is only used in tests.
```

Comment:
Not 100% critical for the proposal. But maybe more curios question ... knowing whether the resource is created or patched is quite important when analyzing different issues (e.g. some other tooling deleting some resources). Is there some way to get the result of the ServerSideApply call at the log level? E.g. be able to for example log at debug something like `Resource X with name Y was created`, `Resource X with name Y was patched`, or `Resource X with name Y did not change`

---

Additionally - and this might be more important - the `ReconcileResult` has 4 states. Created, patched, deleted and noop. Is the noop also not used for anything? I think that in the past we used it to for example see if certificates changed. But I think we might not do that anymore.

Reply:
> Is there some way to get the result of the ServerSideApply call at the log level?

Looks like that's a yes from the API perspective
- https://github.com/kubernetes/kubernetes/blob/56b5f05aea6f32db3b81417ce5d3508dc80a33ea/staging/src/k8s.io/client-go/rest/request.go#L1389-L1394
- https://github.com/kubernetes/apiserver/blob/1f1467cf86675941d0552b8d6baf69abd7df420b/pkg/endpoints/handlers/patch.go#L234-L237

However, I don't see that this is exposed in the fabric8 client

Reply:
Thanks for looking into it. For issue analysis, I guess Fabric8 might log it somewhere (or it should be at least easy to add to it). So that should be fine.

Reply:
Back to the second part abotu the `noop` -> there seem to be at least three places where we depend on it (search for `patchResult instanceof ReconcileResult.Noop` - there might be more with other code, not sure). So I think we should include how to deal with that as well.

Reply:
Obviously this is still a proposal and I am not a Java engineer able to test, but I'd image `.patch()` call would still return a `patchResult` in a [similar way](https://github.com/strimzi/strimzi-kafka-operator/blob/834bfc894475a3204029c2abd1426d4200ca3b22/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaExporterReconciler.java#L191C21-L192C46), I am struggling to find where it is defined in Kubernetes code but I'd suggest this is something that can be handled during the implementation of the proposal.

Reply:
The problem is that it seems we will not have information on whether the result is a patch, create or noop. So this condition will not work and will need to be worked around. We can probably leave it to the PR. But we should not forget about it as it might not necessarily throw an error in the compiler.